app/controllers/actions/IdentifierAction.scala

```diff
@@ -43,181 +44,309 @@ trait IdentifierActionProvider {
+    )(sessionData => {
+      getEnrolmentValueOpt(Enrolment.Agent.key, Enrolment.Agent.value, enrolments) match {
+        case Some(_) =>
+          authorised(predicate(sessionData.mtditid)) {
+            infoLogger("Agent authentication completed successfully for user. Invoking block for request")
+            block(IdentifierRequest(request, sessionData, isAgent = true))
+          }.recoverWith {
+            agentRecovery(sessionData.mtditid)
+          }
+        case None =>
+          warnLogger("Could not find HMRC-AS-AGENT enrolment for user. Redirecting to agent set-up")
+          Future.successful(Redirect(config.setUpAgentServicesAccountUrl()))
       }
-    }
+    })
   }
 
-  def predicate(mtdId: String): Predicate =
-    HMRCEnrolment("HMRC-MTD-IT")
-      .withIdentifier("MTDITID", mtdId)
-      .withDelegatedAuthRule("mtd-it-auth")
+  private def agentRecovery(mtdItId: String)
+                           (implicit hc: HeaderCarrier): PartialFunction[Throwable, Future[Result]] = {
+    val methodLoggingContext: String = "agentRecovery"
 
-  private def secondaryAgentPredicate(mtdId: String): Predicate =
-    HMRCEnrolment("HMRC-MTD-IT-SUPP")
-      .withIdentifier("MTDITID", mtdId)
-      .withDelegatedAuthRule("mtd-it-auth-supp")
+    val warnLogger: String => Unit = warnLog(
+      secondaryContext = methodLoggingContext,
+      extraContext = Some("invokeBlock")
+    )
 
-  private def agentAuth[A](request: Request[A], block: IdentifierRequest[A] => Future[Result], enrolments: Enrolments)
-                          (implicit hc: HeaderCarrier): Future[Result] = {
+    val errorLogger: String => Unit = errorLog(
+      secondaryContext = methodLoggingContext,
+      extraContext = Some("invokeBlock")
+    )
 
-    getSessionData(request).flatMap {
-      case Some(sessionData) =>
-        authorisedAgentServices(enrolments) match {
-          case Some(_) =>
-            authorised(predicate(sessionData.mtditid)){
-              logger.info(s"[AuthorisedAction][agentAuthentication] - authorised as Primary Agent")
-              block(IdentifierRequest(request, sessionData.mtditid, isAgent = true))
-            }.recoverWith{
-              agentRecovery(sessionData.mtditid)
-            }
-          case None =>
-            logger.warn("User did not have HMRC-AS-AGENT enrolment ")
-            Future.successful(Redirect(config.setUpAgentServicesAccountUrl))
+    val agentErrorString = "Agent authentication failed with reason:"
+    val supportingAgentErrorString = "Supporting agent authentication failed with reason:"
+
+    {
+      case _: NoActiveSession =>
+        warnLogger(s"$agentErrorString No active session. Redirecting to V&C")
+        Future(Redirect(config.viewAndChangeEnterUtrUrl))
+      case e: AuthorisationException =>
+        warnLogger(s"$agentErrorString ${e.getMessage}. Checking for supporting agent credentials")
+        authorised(secondaryAgentPredicate(mtdItId)) {
```

Comment:
I believe this is a third call to auth in a row here, we should be able to condense this down by using appropriate retrievals

Reply:
We had a discussion with the auth team about this already. We cannot determine wether the user is a secondary agent via retrievals, we can only do it by attempting to authorise against a secondary agent predicate and seeing if it passes or fails

Reply:
I which case this third call should not be needed since we can combine retrievals like this. 

```
    def agentEnrolment(mtdId: String): Enrolment =
      Enrolment("HMRC-MTD-IT")
        .withIdentifier("MTDITID", mtdId)
        .withDelegatedAuthRule("mtd-it-auth")

    def supportingAgentEnrolment(mtdId: String): Enrolment =
      Enrolment("HMRC-MTD-IT-SUPP")
        .withIdentifier("MTDITID", mtdId)
        .withDelegatedAuthRule("mtd-it-auth-sup")

    def agentPredicate(mtdId: String) = AffinityGroup.Agent and (agentEnrolment(mtdId) or supportingAgentEnrolment(mtdId))
```

This would be more efficient for us and better for Auth. We should be avoiding situations like this where we put this sort of pressure on upstream services.